ext/random/random.c

```diff
@@ -0,0 +1,1856 @@
+{
+	engine_common_free_obj(object);
+}
+
+PHPAPI zend_object *php_random_engine_common_clone_object(zend_object *object)
+{
+	return engine_common_clone_obj(object);
+}
+/* ---- COMMON API END ---- */
+
+/* ---- COMBINED LCG BEGIN ---- */
+/*
+ * combinedLCG() returns a pseudo random number in the range of (0, 1).
+ * The function combines two CGs with periods of
+ * 2^31 - 85 and 2^31 - 249. The period of this function
+ * is equal to the product of both primes.
+ */
+#define MODMULT(a, b, c, m, s) q = s/a;s=b*(s-a*q)-c*q;if(s<0)s+=m
+
+static inline void combinedlcg_seed_default(php_random_status *status)
+{
+	php_random_status_state_combinedlcg *s = status->state;
+	struct timeval tv;
+
+	if (gettimeofday(&tv, NULL) == 0) {
+		s->state[0] = tv.tv_usec ^ (tv.tv_usec << 11);
+	} else {
+		s->state[0] = 1;
+	}
+
+#ifdef ZTS
+	s->state[1] = (zend_long) tsrm_thread_id();
+#else
+	s->state[1] = (zend_long) getpid();
+#endif
+
+	/* Add entropy to s2 by calling gettimeofday() again */
+	if (gettimeofday(&tv, NULL) == 0) {
+		s->state[1] ^= (tv.tv_usec << 11);
+	}
+}
+
+static void combinedlcg_seed(php_random_status *status, uint64_t seed)
+{
+	php_random_status_state_combinedlcg *s = status->state;
+
+	s->state[0] = seed & 0xffffffffU;
+	s->state[1] = seed >> 32;
+}
+
+static uint64_t combinedlcg_generate(php_random_status *status)
+{
+	php_random_status_state_combinedlcg *s = status->state;
+	int32_t q, z;
+
+	MODMULT(53668, 40014, 12211, 2147483563L, s->state[0]);
+	MODMULT(52774, 40692, 3791, 2147483399L, s->state[1]);
+
+	z = s->state[0] - s->state[1];
+	if (z < 1) {
+		z += 2147483562;
+	}
+
+	return (uint64_t) z;
+}
+
+static zend_long combinedlcg_range(php_random_status *status, zend_long min, zend_long max)
+{
+	return common_range(&php_random_algo_combinedlcg, status, min, max);
+}
+
+static bool combinedlcg_serialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_combinedlcg *s = status->state;
+	zval t;
+	uint32_t i;
+
+	for (i = 0; i < 2; i++) {
+		ZVAL_LONG(&t, s->state[i]);
+		zend_hash_next_index_insert(data, &t);
+	}
+
+	return true;
+}
+
+static bool combinedlcg_unserialize(php_random_status *status, HashTable *data)
+{
+	php_random_status_state_combinedlcg *s = status->state;
+	zval *t;
+	uint32_t i;
+
+	for (i = 0; i < 2; i++) {
+		t = zend_hash_index_find(data, i);
+		if (!t || Z_TYPE_P(t) != IS_LONG) {
+			return false;
+		}
+		s->state[i] = Z_LVAL_P(t);
+	}
+
+	return true;
+}
+
+const php_random_algo php_random_algo_combinedlcg = {
+	sizeof(uint32_t),
+	sizeof(php_random_status_state_combinedlcg),
+	combinedlcg_seed,
+	combinedlcg_generate,
+	combinedlcg_range,
+	combinedlcg_serialize,
+	combinedlcg_unserialize
+};
+
+static zend_object *php_random_engine_combinedlcg_new(zend_class_entry *ce)
+{
+	return &php_random_engine_common_init(ce,&random_engine_combinedlcg_object_handlers, &php_random_algo_combinedlcg)->std;
+}
+/* ---- COMBINED LCG END ---- */
+
+/* ---- MT BEGIN ---- */
```

Comment:
Would it make sense to split `random.c` into a dedicated C file per engine? Each engine comes with quite a few functions and I find `random.c` hard to grok just with the comments in between.

Reply:
That is correct.
However, at one time the Engine was repeatedly increasing and decreasing, and it was very tedious to modify the build scripts each time.
As soon as the final implementation algorithm is finalized, we will cut them out as separate files.

Reply:
Updated TODO to include an item separating engine implementation.